Added Full Support for Reminders.

YBTopaz8 opened the pull request:
- Users Can Add/Update/Remove a reminder from event
- Should be robust to handle errors
- Tried to decouple so that we can later expose those methods for granular dev control.
- Also updated the sample for easy test.

Fixes #52

YBTopaz8 committed:
Added Full Support for Reminders.
- Users Can Add/Update/Remove a reminder from event
- Should be robust to handle errors
- Tried to decouple so that we can later expose those methods for granular dev control.

YBTopaz8 commented:
@jfversluis Should be good now.
I intend to add Tasks Support too but down the line (coming weeks)
Let me know !

jfversluis commented:
Thank you so much for this! I'll have a look when I can!

jfversluis commented:
Mostly love it! Couple of things:

* No implementation for iOS/macOS & Windows?
* Should we consider adding a Alarm object or something and allow people to add multiple? I think all platforms support that?
* For the sample: why can't I add a reminder for a full-day event?

YBTopaz8 commented:
> Mostly love it! Couple of things:
> 
> * No implementation for iOS/macOS & Windows?
> * Should we consider adding a Alarm object or something and allow people to add multiple? I think all platforms support that?
> * For the sample: why can't I add a reminder for a full-day event?

Thank you!

So
1. Sadly no for MacOS since I do not really own any such devices and for windows, I tried but I faced some issues, please have a look at my "analysis" [here](https://github.com/jfversluis/Plugin.Maui.CalendarStore/issues/52#issuecomment-2491698716) and let me know perhaps!

2. Perhaps, yes! And yes we can add several alarms for an event, I can add that, yes! (For Android)

3. Because I added 
```
void HasReminderChkbx_CheckedChanged(object sender, CheckedChangedEventArgs e)
	{
		IsAllDaySection.IsVisible=!e.Value;
	}
```
 in `AddEventsPage.xaml.cs`
Please remove it and it should be good! A day can have several (overlapping) events alarms/reminders.

jfversluis committed:
Add Reminders

jfversluis commented:
OK I pushed some changes implementing this for iOS.

I didn't touch Android, if you could update that, that would be great :) 

Also I didn't update the sample yet to support this. If we can get this for Android then I will have a look at Windows.

YBTopaz8 commented:
Very well, I'll add the equivalents for Android in a few

jfversluis committed:
Implement Windows & minor fixes

jfversluis committed:
Minor improvements

jfversluis commented:
Added the Windows implementation and made stuff build. I will probably also look into adding this to the sample app so you can focus on the Android implementation!

YBTopaz8 committed:
This SHOULD solve issues with null calendar titles being null

YBTopaz8 committed:
Added support for Multiple reminders in a day
- Added a RemindersInMinutes Property to ease the use of minutes directly.

YBTopaz8 committed:
Should be good now.

YBTopaz8 committed:
Merged

YBTopaz8 commented:
@jfversluis let me know if anything.
I exposed a ReminderInMinutes prop to make it easier for use.

jfversluis commented:
Looks like something didn't merge correctly because I can't compile `CalendarStore.android.cs` around line 780

YBTopaz8 committed:
Added Missing Code parts.

Fixed the demo and it now shows full CRUD with reminders

I had all my projects mixed up - sorry!

YBTopaz8 commented:
All good now @jfversluis . Apologies for the mix up - had so much with little time but used opportunity to fix the demo and now we can do full CRUD.

Please tell me know of anything if anything.

jfversluis commented on .gitignore:
```diff
@@ -397,3 +397,4 @@ FodyWeavers.xsd
 # JetBrains Rider
 *.sln.iml
 .DS_Store
+/src/Plugin.Maui.CalendarStore/CalendarStore.windows.cs
```
This doesn't seem right, did I do that? 😅

YBTopaz8 replied:
Perhaps!
I use VS only, I don't think I made the changes  (unless some weird bug happened and I don't know

YBTopaz8 replied:
I may confirm, the FodyWeavers.xsd file isn't from me (so far after checking!)

jfversluis replied:
Its not about the xsd file, its `/src/Plugin.Maui.CalendarStore/CalendarStore.windows.cs` but I removed it for you

jfversluis commented on samples/Plugin.Maui.CalendarStore.Sample/AddEventsPage.xaml.cs:
```diff
@@ -52,10 +56,15 @@ protected override async void OnNavigatedTo(NavigatedToEventArgs args)
 			EventIsAllDay = eventToUpdate.IsAllDay;
-
+			EventReminders = new ObservableCollection<Reminder>(eventToUpdate.Reminders);
+			//MinsBeforeReminder = eventToUpdate.MinutesBeforeReminder;
```
Lets remove commented line

jfversluis commented on samples/Plugin.Maui.CalendarStore.Sample/AddEventsPage.xaml.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using System.Collections.ObjectModel;
+using System.Diagnostics;
```
Is this needed?

YBTopaz8 replied:
To print `Debug.WriteLine()` 

Removed..

jfversluis commented on src/Plugin.Maui.CalendarStore/CalendarEvent.shared.cs:
```diff
@@ -46,26 +46,39 @@ public CalendarEvent(string id, string calendarId, string title)
+	/// Gets whether this event has at least one reminder.
+	/// </summary>
+	public bool HasReminders => Reminders?.Count > 0;
+	public int MinutesBeforeReminder { get; internal set; }
```
Why do we need this? It only works on Android now and if we have multiple reminders, which one is this for?

YBTopaz8 replied:
I figured it it would help to quickly have a prop that reports that?

Removed...

jfversluis commented on src/Plugin.Maui.CalendarStore/CalendarStore.android.cs:
```diff
@@ -527,15 +781,30 @@ IEnumerable<CalendarEvent> ToEvents(ICursor cur, List<string> projection)
 		var end = DateTimeOffset.FromUnixTimeMilliseconds(cursor.GetLong(projection.IndexOf(CalendarContract.Events.InterfaceConsts.Dtend)));
 
-		return new(cursor.GetString(projection.IndexOf(CalendarContract.Events.InterfaceConsts.Id)) ?? string.Empty,
+		DateTimeOffset SafeConvertTime(DateTimeOffset time, string tz)
```
This seems useful, but the method is never used?

YBTopaz8 replied:
To Event is used by `ToEvents`() and  `GetEvent()` so are needed. Please have a look at lines 224 and 775 in `CalendarStore.android.cs`
You should see `return ToEvent(cursor, eventsColumns);` and  `yield return ToEvent(cur, projection);` respectively

jfversluis commented on src/Plugin.Maui.CalendarStore/CalendarStore.android.cs:
```diff
@@ -377,12 +389,252 @@ public async Task UpdateEvent(string eventId, string title, string description,
+		}
+	}
+
+	void AddSingleReminder(long eventId, int reminderMinutes)
```
I don't see why we need separate calls for single and all/multiple reminders? Probably has to do with that minutes to reminder field. Just remove that and let people add reminders through the `Reminder` object. 

We can think about maybe making easier helper calls that will just add 1 reminder by specifying some minutes in a follow up PR.

YBTopaz8 replied:
Removed AddSingleReminder.

Future PRs may indeed add more.

jfversluis commented on src/Plugin.Maui.CalendarStore/CalendarStore.android.cs:
```diff
@@ -377,12 +389,252 @@ public async Task UpdateEvent(string eventId, string title, string description,
+			eventToUpdate.Location, eventToUpdate.StartDate, eventToUpdate.EndDate, eventToUpdate.IsAllDay, eventToUpdate.Reminders.ToArray());
+
+	/// <inheritdoc/>
+	void AddMultipleReminders(long eventId, Reminder[]? reminders)
```
Let's name this `AddReminders`, the multiple is already implied by the fact that it takes an array.

Also `/// <inheritdoc/>` in this case doesn't do anything

YBTopaz8 replied:
Done.

YBTopaz8 committed:
Applied suggested changes/

Should Hopefully Align to the PR rules?

jfversluis committed:
Update .gitignore

jfversluis committed:
Fix all the things

jfversluis committed:
Merge branch 'main' into pr/55

jfversluis commented:
Cleaned up a bunch of code and improved the sample app, I think this is ready to go now!
